### ISSUE TITLE
Update watchers as VFS is updated

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#11902]
Instead of registering watchers at the end of the build we could register them as part of VFS operations. I.e. when directories are deleted we can stop watching them, and when new content is discovered we can start watching it.

This would also speed up builds as watcher registration wouldn't have to happen at the end of the build, and would happen largely in parallel.

This would require a fully event-driven interface for watchers (see #11813 for Linux support).

See https://github.com/gradle/gradle/issues/11902 about keeping watchers alive during the build.

---
cc: @gradle/build-cache

**Comment by wolfs** [comment]
Duplicate of #12437.
